runtime/src/main/java/org/corfudb/runtime/collections/SMRMultiIndex.java

```diff
@@ -0,0 +1,362 @@
+     *
+     * @param index             The previous state of the index
+     * @param key               The key from the put call
+     * @param value             The value from the put call. This is not
+     *                          needed to generate an undo record.
+     * @return                  An undo record, which for a put is the
+     *                          previous value in the map.
+     */
+    @DontInstrument
+    V undoPutRecord(SMRMultiIndex<K,V,I,P> index, K key, V value) {
+        return index.get(key);
+    }
+
+    /** Undo a put, given the current state of the map, an undo record
+     * and the arguments to the put command to undo.
+     *
+     * @param index           The state of the map after the put to undo
+     * @param undoRecord    The undo record generated by undoPutRecord
+     * @param key           The key of the put to undo
+     * @param value         The value of the put to undo, which is not
+     *                      needed.
+     */
+    @DontInstrument
+    void undoPut(SMRMultiIndex<K,V,I,P> index, V undoRecord, K key, V value) {
+        if (undoRecord == null) {
+            index.remove(key);
+        } else {
+            index.put(key, undoRecord);
+        }
+    }
+
+    @Accessor
+    @Override
+    public V get(Object key) {
+        return mainMap.get(key);
+    }
+
+
+
+    @Override
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove", undoRecordFunction = "undoRemoveRecord")
+    //@MutatorAccessor(name = "remove")
+    public V remove(Object key) {
+        V previous = mainMap.remove(key);
+        if(previous != null)
+            removeIndexes((K)key, previous);
+        return previous;
+    }
+
+    /** Generate an undo record for a remove, given the previous state of the map
+     * and the parameters to the remove call.
+     *
+     * @param index             The previous state of the map
+     * @param key               The key from the remove call
+     * @return                  An undo record, which for a remove is the
+     *                          previous value in the map.
+     */
+    @DontInstrument
+    V undoRemoveRecord(SMRMultiIndex<K,V,I,P> index, K key) {
+        return index.get(key);
+    }
+
+    /** Undo a remove, given the current state of the map, an undo record
+     * and the arguments to the remove command to undo.
+     *
+     * @param index           The state of the map after the put to undo
+     * @param undoRecord    The undo record generated by undoRemoveRecord
+     */
+    @DontInstrument
+    void undoRemove(SMRMultiIndex<K,V,I,P> index, V undoRecord, K key) {
+        if (undoRecord == null) {
+            index.remove(key);
+        } else {
+            index.put(key, undoRecord);
+        }
+    }
+
+    @Mutator(name = "putAll")
+    @Override
+    public void putAll(Map<? extends K, ? extends V> m) {
+
+        throw new RuntimeException("###########Operation Not Implemented.##########");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)